### PR TITLE
topologically sort contract info in dex module

### PR DESCRIPTION
**Pull request by codchen**
- Allow contracts to specify dependencies during contract registration
- Allow contract to update its registered information
- Reject registration if the contract would introduce a circular dependency
- Topologically sort contracts to decide processing order in `dex` module Begin/EndBlock

### CODE REVIEW

File: x/dex/contract/dag.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/sei-protocol/sei-chain/utils"
+	"github.com/sei-protocol/sei-chain/x/dex/types"
+)
+
+type node struct {
+	contractAddr  string
+	incomingNodes *utils.StringSet
+}
+
+// Kahn's algorithm
```
Comment (philipsu522) [inline x/dex/contract/dag.go:13]: wonder if this can be generalized for module sorting. Can be addressed in a new pr though

**Review by philipsu522 — COMMENTED** [review]
🔥

**Review by philipsu522 — APPROVED** [review]
🔥

File: x/dex/contract/dag.go (hunk 1 above)
Comment (codchen) [reply x/dex/contract/dag.go:13]: yeah that should be possible. A bit sad that golang doesn't support generics so it might get a little verbose.

**Review by LCyson — APPROVED** [review]

**Pull request merged**